File: src/bun/updates.ts
import { Updater } from "electrobun/main";
import { notify } from "../core/notify";

/** How often to look for a new release in the background. */
const CHECK_EVERY_MS = 24 * 3600_000;

export type UpdatePhase = "idle" | "checking" | "downloading" | "ready" | "error";

/**
 * Checks GitHub Releases (see `release.baseUrl` in electrobun.config.ts) and downloads a
 * new version in the background. Installing waits for the user, since it restarts the app.
 */
export class Updates {
  phase: UpdatePhase = "idle";
  version = "";
  error = "";
  private lastCheckAt = 0;

  constructor(private onChange: () => void = () => {}) {}

  /** Check shortly after launch, then once a day. Hourly ticks compare wall-clock time, so sleep doesn't delay it. */
  run() {
    setTimeout(() => void this.check(false), 60_000);
    setInterval(() => {
      if (Date.now() - this.lastCheckAt >= CHECK_EVERY_MS) void this.check(false);
    }, 3600_000);
  }

  get busy() {
    return this.phase === "checking" || this.phase === "downloading";
  }

  /** Look for a new version and download it. `manual` also reports "up to date" and errors. */
  async check(manual: boolean) {
    if (this.busy || this.phase === "ready") return;
    this.lastCheckAt = Date.now();
    this.set("checking");
    try {
      let info = await Updater.checkForUpdate();
      if (info.error) throw new Error(info.error);
      if (!info.updateAvailable) {
        this.set("idle");
        if (manual) notify("Usage Window Starter", `You're up to date (${await Updater.localInfo.version()})`);
        return;
      }
      this.version = info.version;
      if (!info.updateReady) {
        this.set("downloading");
        await Updater.downloadUpdate();
        info = Updater.updateInfo();
        if (info.error || !info.updateReady) throw new Error(info.error || "download didn't finish");
      }
      this.set("ready");
      notify("Usage Window Starter", `Version ${this.version} is ready. Choose "Install update" in the menu.`);
    } catch (err) {
      this.error = err instanceof Error ? err.message : String(err);
      this.set("error");
      if (manual) notify("Usage Window Starter", `Update check failed: ${this.error.slice(0, 120)}`);
    }
  }

  /** Replace the app with the downloaded version and restart. Quits on success. */
  async install() {
    if (this.phase !== "ready") return;
    await Updater.applyUpdate();
    // Still running means it failed.
    this.error = Updater.updateInfo().error || "install failed";
    this.set("error");
    notify("Usage Window Starter", `Couldn't install the update: ${this.error.slice(0, 120)}`);
  }

  private set(phase: UpdatePhase) {
    this.phase = phase;
    if (phase !== "error") this.error = "";
    this.onChange();
  }
}

File: src/go/updates.go
package main

import (
	"context"
	"fmt"
	"path/filepath"
	"sync"
	"time"

	"electrobun"

	"usagewindowstarter/src/go/core"
	"usagewindowstarter/src/go/updater"
)

// checkEvery is how often to look for a new release in the background.
const checkEvery = 24 * time.Hour

// Updates checks GitHub Releases (release.baseUrl in electrobun.config.ts) and
// downloads a new version in the background. Installing waits for the user, since
// it restarts the app.
type Updates struct {
	mu        sync.Mutex
	u         *updater.Updater
	phase     string // idle, checking, downloading, ready, installing, error
	version   string
	err       string
	lastCheck time.Time
	onChange  func()
}

func newUpdates(bundle electrobun.BundlePaths, onChange func()) *Updates {
	u, err := updater.New(bundle.ExeDir, bundle.ResourcesDir)
	up := &Updates{u: u, phase: "idle", onChange: onChange}
	if err != nil {
		up.err = err.Error()
	} else {
		core.AppVersion = u.Info.Version
		u.HelperLog = filepath.Join(core.HomeDir, "update-helper.log")
	}
	return up
}

func (up *Updates) status() (phase, version string) {
	up.mu.Lock()
	defer up.mu.Unlock()
	return up.phase, up.version
}

func (up *Updates) set(phase string) {
	up.mu.Lock()
	up.phase = phase
	if phase != "error" {
		up.err = ""
	}
	up.mu.Unlock()
	up.onChange()
}

// run checks shortly after launch, then once a day. Hourly ticks compare
// wall-clock time, so sleep doesn't delay it.
func (up *Updates) run() {
	time.AfterFunc(3*time.Second, up.reportLastInstall)
	time.AfterFunc(time.Minute, func() { up.check(false) })
	go func() {
		for range time.Tick(time.Hour) {
			up.mu.Lock()
			due := time.Since(up.lastCheck) >= checkEvery
			up.mu.Unlock()
			if due {
				up.check(false)
			}
		}
	}()
}

// reportLastInstall tells the user how the last install went. The helper does its
// work after this app quits, so a failure is only known on the next launch.
func (up *Updates) reportLastInstall() {
	if up.u == nil {
		return
	}
	r, err := up.u.NewResult()
	if err != nil || r == nil {
		return
	}
	if r.Success {
		core.Log("event", "update-complete", "version", r.Version)
		core.Notify(appName, "Updated to version "+r.Version)
		return
	}
	core.Log("event", "update-error", "phase", "native-"+r.Phase, "version", r.Version, "error", r.Message)
	core.Notify(appName, fmt.Sprintf("Update to %s failed during %s: %s", r.Version, r.Phase, truncate(r.Message, 100)))
}

// check looks for a new version and downloads it. manual also reports
// "up to date" and errors.
func (up *Updates) check(manual bool) {
	// Claim the check under the lock: panel actions arrive on separate goroutines, and
	// two checks at once would share the manifest and the download files.
	up.mu.Lock()
	if up.phase == "checking" || up.phase == "downloading" || up.phase == "ready" || up.phase == "installing" {
		up.mu.Unlock()
		return
	}
	up.phase, up.err = "checking", ""
	up.lastCheck = time.Now()
	up.mu.Unlock()
	up.onChange()
	current := ""
	if up.u != nil {
		current = up.u.Info.Version
	}
	phase := "check"
	fail := func(err error) {
		core.Log("event", "update-error", "phase", phase, "manual", manual, "error", err.Error())
		up.mu.Lock()
		up.err = err.Error()
		up.mu.Unlock()
		up.set("error")
		if manual {
			core.Notify(appName, "Update check failed: "+truncate(err.Error(), 120))
		}
	}
	if up.u == nil {
		fail(fmt.Errorf("can't read version.json"))
		return
	}

	res, err := up.u.Check(context.Background())
	if err != nil {
		fail(err)
		return
	}
	core.Log("event", "update-check", "manual", manual, "current", current, "latest", res.Version,
		"available", res.Available, "alreadyDownloaded", res.Ready)
	if !res.Available {
		up.set("idle")
		if manual {
			core.Notify(appName, fmt.Sprintf("You're up to date (%s)", up.u.Info.Version))
		}
		return
	}
	up.mu.Lock()
	up.version = res.Version
	up.mu.Unlock()
	if !res.Ready {
		phase = "download"
		up.set("downloading")
		started := time.Now()
		if err := up.u.Download(context.Background()); err != nil {
			fail(err)
			return
		}
		core.Log("event", "update-download", "version", res.Version, "seconds", int(time.Since(started).Seconds()))
	}
	up.set("ready")
	core.Notify(appName, fmt.Sprintf(`Version %s is ready. Choose "Install" in the panel.`, res.Version))
}

// install replaces the app with the downloaded version and restarts it.
func (up *Updates) install() {
	// Claim the install under the lock: duplicate clicks arrive on separate goroutines
	// and must not start two helpers against the same bundle.
	up.mu.Lock()
	if up.phase != "ready" {
		up.mu.Unlock()
		return
	}
	up.phase = "installing"
	version := up.version
	up.mu.Unlock()
	up.onChange()
	core.Log("event", "update-install", "from", up.u.Info.Version, "to", version)
	app.scheduler.Stop()
	if err := up.u.Apply(); err != nil {
		core.Log("event", "update-error", "phase", "install", "error", err.Error())
		up.mu.Lock()
		up.err = err.Error()
		up.mu.Unlock()
		up.set("error")
		core.Notify(appName, "Couldn't install the update: "+truncate(err.Error(), 120))
		app.scheduler.Run()
		return
	}
	quit() // the helper waits for this app to exit, then swaps it and relaunches
}

func truncate(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}
